src/main/resources/assets/admin/common/js/ui/treegrid/TreeGrid.ts

```diff
@@ -627,231 +581,91 @@ export class TreeGrid<DATA>
-
-        this.setActive(false);
-
-        this.grid.invalidate();
-
-        root.setExpanded(true);
-        this.initData(root.treeToList());
-        this.invalidate();
-
-        this.setActive(true);
+    reload(): Q.Promise<void> {
+        this.root.resetCurrentRoot();
+        this.gridData.setItems([], this.idPropertyName);
 
-        this.notifyLoaded();
+        return this.doExpandNode(this.root.getCurrentRoot()).then(() => {
+            if (this.hasHighlightedNode()) {
+                this.highlightCurrentNode();
+            }
+        }).catch((reason: any) => {
+            this.initData([]);
+            this.handleError(reason);
+        }).then(() => {
+            this.setActive(true);
+            this.notifyLoaded();
+        });
     }
 
-    updateNode(data: DATA, oldDataId?: string): Q.Promise<void> {
-
-        let dataId = oldDataId || this.getDataId(data);
-        let nodeToUpdate = this.root.getCurrentRoot().findNode(dataId);
-
-        if (!nodeToUpdate) {
-            throw new Error('TreeNode to update not found: ' + dataId);
-        }
-
-        return this.fetchAndUpdateNodes([nodeToUpdate], oldDataId ? this.getDataId(data) : undefined);
+    updateNodesByData(dataItems: DATA[]) {
+        dataItems.forEach((updatedData: DATA) => {
+            this.updateNodeByData(updatedData);
+        });
     }
 
-    updateNodeByDataId(dataId: string): Q.Promise<void> {
-        const nodesToUpdate: TreeNode<DATA>[] = this.root.getNodesByDataId(dataId);
-
-        if (nodesToUpdate.length > 0) {
-            return this.fetchAndUpdateNodes(nodesToUpdate);
-        }
+    updateNodeByData(updatedData: DATA, oldDataId?: string) {
+        const dataId: string = !!oldDataId ? oldDataId : updatedData.getId();
+        const treeNodes: TreeNode<DATA>[] = this.root.getNodesByDataId(dataId);
 
-        return Q(null);
+        treeNodes.forEach((updatedNode: TreeNode<DATA>) => {
+            updatedNode.setData(updatedData);
+            this.invalidateNodes([updatedNode]);
+            this.notifyDataChanged(new DataChangedEvent<DATA>([updatedNode], DataChangedType.UPDATED));
+        });
     }
 
-    updateNodes(data: DATA, oldDataId?: string): Q.Promise<void> {
+    deleteNodeByDataId(dataId: string) {
+        const nodes: TreeNode<DATA>[] = this.root.getNodesByDataId(dataId);
 
-        let dataId = oldDataId || this.getDataId(data);
-        let nodesToUpdate = this.root.getCurrentRoot().findNodes(dataId);
-
-        if (!nodesToUpdate) {
-            throw new Error('TreeNode to update not found: ' + dataId);
-        }
-
-        return this.fetchAndUpdateNodes(nodesToUpdate, oldDataId ? this.getDataId(data) : undefined);
+        nodes.forEach((node: TreeNode<DATA>) => {
+            this.deleteNode(node);
+        });
     }
 
-    deleteNode(data: DATA): void {
-        if (this.highlightedNode && this.highlightedNode.getDataId() === this.getDataId(data)) {
+    protected deleteNode(node: TreeNode<DATA>): void {
```

Comment:
Codacy found an issue: [Method 'deleteNode' has a complexity of 7. Maximum allowed is 4.](https://app.codacy.com/manual/enonic/lib-admin-ui/pullRequest?prid=6401650)